// egg-bin/dist/middleware/global_options.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_util_1 = require("node:util");
const node_path_1 = __importDefault(require("node:path"));
const artus_cli_1 = require("@artus-cli/artus-cli");
const utils_1 = require("../utils");
const debug = (0, node_util_1.debuglog)('egg-bin:midddleware:global_options');
let default_1 = class default_1 {
    program;
    async configDidLoad() {
        // add global options
        this.program.option({
            base: {
                description: 'directory of application, default to `process.cwd()`',
                type: 'string',
                alias: 'baseDir',
            },
            typescript: {
                description: 'whether enable typescript support',
                type: 'boolean',
                alias: 'ts',
            },
            tscompiler: {
                description: 'ts compiler, like ts-node/register, ts-node/register/transpile-only, @swc-node/register, esbuild-register etc',
                type: 'string',
                alias: 'tsc',
            },
        });
        this.program.use(async (ctx, next) => {
            debug('before next');
            if (!ctx.args.base) {
                ctx.args.base = ctx.cwd;
                debug('ctx.args.base not set, auto set it to cwd: %o', ctx.cwd);
            }
            if (!node_path_1.default.isAbsolute(ctx.args.base)) {
                ctx.args.base = node_path_1.default.join(ctx.cwd, ctx.args.base);
            }
            debug('matched cmd: %o, ctx.args.base: %o', ctx.matched?.cmd, ctx.args.base);
            const pkg = await (0, utils_1.readPackageJSON)(ctx.args.base);
            ctx.args.pkgEgg = pkg.egg ?? {};
            const tscompiler = ctx.args.tscompiler ?? ctx.env.TS_COMPILER ?? ctx.args.pkgEgg.tscompiler;
            if (ctx.args.typescript === undefined) {
                // try to ready EGG_TYPESCRIPT env first, only accept 'true' or 'false' string
                if (ctx.env.EGG_TYPESCRIPT === 'false') {
                    ctx.args.typescript = false;
                    debug('detect typescript=%o from EGG_TYPESCRIPT=%o', false, ctx.env.EGG_TYPESCRIPT);
                }
                else if (ctx.env.EGG_TYPESCRIPT === 'true') {
                    ctx.args.typescript = true;
                    debug('detect typescript=%o from EGG_TYPESCRIPT=%o', true, ctx.env.EGG_TYPESCRIPT);
                }
                else if (typeof ctx.args.pkgEgg.typescript === 'boolean') {
                    // read `egg.typescript` from package.json if not pass argv
                    ctx.args.typescript = ctx.args.pkgEgg.typescript;
                    debug('detect typescript=%o from pkg.egg.typescript=%o', true, ctx.args.pkgEgg.typescript);
                }
                else if (pkg.dependencies?.typescript) {
                    // auto detect pkg.dependencies.typescript or pkg.devDependencies.typescript
                    ctx.args.typescript = true;
                    debug('detect typescript=%o from pkg.dependencies.typescript=%o', true, pkg.dependencies.typescript);
                }
                else if (pkg.devDependencies?.typescript) {
                    ctx.args.typescript = true;
                    debug('detect typescript=%o from pkg.devDependencies.typescript=%o', true, pkg.devDependencies.typescript);
                }
                else if (await (0, utils_1.hasTsConfig)(ctx.args.base)) {
                    // tsconfig.json exists
                    ctx.args.typescript = true;
                    debug('detect typescript=%o cause tsconfig.json exists', true);
                }
                else if (tscompiler) {
                    ctx.args.typescript = true;
                    debug('detect typescript=%o from --tscompiler=%o', true, tscompiler);
                }
            }
            if (ctx.args.typescript) {
                const findPaths = [node_path_1.default.dirname(__dirname)];
                if (tscompiler) {
                    // try app baseDir first on custom tscompiler
                    findPaths.unshift(ctx.args.base);
                }
                ctx.args.tscompiler = tscompiler ?? 'ts-node/register';
                const tsNodeRegister = require.resolve(ctx.args.tscompiler, {
                    paths: findPaths,
                });
                // should require tsNodeRegister on current process, let it can require *.ts files
                // e.g.: dev command will execute egg loader to find configs and plugins
                require(tsNodeRegister);
                // let child process auto require ts-node too
                (0, utils_1.addNodeOptionsToEnv)(`--require ${tsNodeRegister}`, ctx.env);
                // tell egg loader to load ts file
                // see https://github.com/eggjs/egg-core/blob/master/lib/loader/egg_loader.js#L443
                ctx.env.EGG_TYPESCRIPT = 'true';
                // set current process.env.EGG_TYPESCRIPT too
                process.env.EGG_TYPESCRIPT = 'true';
                // load files from tsconfig on startup
                ctx.env.TS_NODE_FILES = process.env.TS_NODE_FILES ?? 'true';
                // keep same logic with egg-core, test cmd load files need it
                // see https://github.com/eggjs/egg-core/blob/master/lib/loader/egg_loader.js#L49
                (0, utils_1.addNodeOptionsToEnv)(`--require ${require.resolve('tsconfig-paths/register')}`, ctx.env);
                debug('set NODE_OPTIONS: %o', ctx.env.NODE_OPTIONS);
            }
            debug('ctx.args: %o', ctx.args);
            debug('enter next');
            await next();
            debug('after next');
        });
    }
};
__decorate([
    (0, artus_cli_1.Inject)(),
    __metadata("design:type", artus_cli_1.Program)
], default_1.prototype, "program", void 0);
__decorate([
    (0, artus_cli_1.LifecycleHook)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], default_1.prototype, "configDidLoad", null);
default_1 = __decorate([
    (0, artus_cli_1.LifecycleHookUnit)()
], default_1);
exports.default = default_1;
